crates/dpapi-transport/src/connect_options.rs

```diff
@@ -0,0 +1,125 @@
+use std::pin::Pin;
+
+use thiserror::Error;
+use url::Url;
+use uuid::Uuid;
+
+/// Type that represents a function for obtaining the sesion token.
+///
+/// We need it because we don't know the destination address in advance.
+///
+/// Parameters:
+/// * `Uuid` is the session id.
+/// * `Url` is the destination of the proxied connection.
+pub type GetSessionTokenFn = dyn Fn(Uuid, Url) -> Pin<Box<dyn Future<Output = std::io::Result<String>>>>;
```

Comment:
As you can notice, I decided to use dynamic dispatch here. There are a few reasons behind this decision:

1. API simplicity.
2. Usage convenience. If I had chosen generics over trait object, then the user would have to specify `GetSessionTokenFn` type even when they do not plan to tunnel the connection.